<a href="https://colab.research.google.com/github/botkin92/Data_analysis_algorithms/blob/lesson_8/Lesson_8_HW.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Урок 8. Снижение размерности данных

In [ ]:
import numpy as np
from sklearn import datasets
from sklearn.model_selection import train_test_split

import xgboost
from sklearn.decomposition import PCA

from sklearn.metrics import accuracy_score, precision_score
# from sklearn.metrics import RocCurveDisplay, auc, roc_curve, roc_auc_score
# import matplotlib.pyplot as plt

1. Обучить любую модель классификации на датасете IRIS до применения PCA и после него. Сравнить качество классификации по отложенной выборке.

In [ ]:
# Загрузим IRIS датасет из sklearn
iris = datasets.load_iris()
X = iris.data
y = iris.target

In [ ]:
n_classes = len(np.unique(y))

In [ ]:
X_train, X_test,y_train, y_test = train_test_split(X, y, test_size=0.5)

In [ ]:
# Для задачи классификации используем xgboost:
model = xgboost.XGBClassifier(objective='multi:softmax', num_classes=n_classes)
model.fit(X_train, y_train)

XGBClassifier(num_classes=3, objective='multi:softprob')

In [ ]:
y_pred = model.predict(X_test)

In [ ]:
accuracy = accuracy_score(y_test, y_pred)
print(f'accuracy:  {accuracy:.3f}')

accuracy:  0.933


In [ ]:
# Разложим датасет на 2 компоненты методом PCA
pca = PCA(n_components=2)
X_train_pca = pca.fit_transform(X_train)
X_test_pca = pca.transform(X_test)

In [ ]:
# Также обучим модель, после разложения
model_pca = xgboost.XGBClassifier(objective='multi:softmax', num_classes=n_classes)
model_pca.fit(X_train_pca, y_train)

XGBClassifier(num_classes=3, objective='multi:softprob')

In [ ]:
y_pred_pca = model_pca.predict(X_test_pca)

In [ ]:
accuracy_pca = accuracy_score(y_test, y_pred_pca)
print(f'accuracy:  {accuracy_pca:.3f}')

accuracy:  0.947


Стоит отметиь, что модель xgboost показала весьма хороший результат - точность предсказания составила 0.933. После применения метода главных компонент результат улучшился до 0.947. Такая метрика (после снижения размерности данных) свидетельствует о сохранении качества данных.

2. Написать свою реализацию метода главных компонент с помощью сингулярного разложения с использованием функции numpy.linalg.svd()